Mathlib/Algebra/Homology/Embedding/Basic.lean

```diff
@@ -0,0 +1,159 @@
+It shall also be used in the construction of the canonical t-structure on the derived
+category of an abelian category (TODO).
+
+## TODO
+
+Define the following:
+- the extension functor `e.extendFunctor C : HomologicalComplex C c ⥤ HomologicalComplex C c'`
+(extending by the zero object outside of the image of `e.f`);
+- assuming `e.IsRelIff`, the restriction functor
+`e.restrictionFunctor C : HomologicalComplex C c' ⥤ HomologicalComplex C c`;
+- the stupid truncation functor
+`e.stupidTruncFunctor C : HomologicalComplex C c' ⥤ HomologicalComplex C c'` which is
+the composition of the two previous functors.
+- assuming `e.IsTruncGE`, truncation functors
+`e.truncGE'Functor C : HomologicalComplex C c' ⥤ HomologicalComplex C c` and
+`e.truncGEFunctor C : HomologicalComplex C c' ⥤ HomologicalComplex C c'`, and a natural
+transformation `e.πTruncGENatTrans : 𝟭 _ ⟶ e.truncGEFunctor C` which is a quasi-isomorphism
+in degrees in the image of `e.f`;
+- assuming `e.IsTruncLE`, truncation functors
+`e.truncLE'Functor C : HomologicalComplex C c' ⥤ HomologicalComplex C c` and
+`e.truncLEFunctor C : HomologicalComplex C c' ⥤ HomologicalComplex C c'`, and a natural
+transformation `e.ιTruncLENatTrans : e.truncGEFunctor C ⟶ 𝟭 _` which is a quasi-isomorphism
+in degrees in the image of `e.f`;
+
+-/
+
+variable {ι ι' : Type*} (c : ComplexShape ι) (c' : ComplexShape ι')
+
+namespace ComplexShape
+
+/-- An embedding of a complex shape `c : ComplexShape ι` into a complex shape
+`c' : ComplexShape ι'` consists of a injective map `f : ι → ι'` which satisfies
+a compatiblity with respect to the relations `c.Rel` and `c'.Rel`. -/
+structure Embedding where
+  /-- the map between the underlying types of indices -/
+  f : ι → ι'
+  injective_f : Function.Injective f
+  rel {i₁ i₂ : ι} (h : c.Rel i₁ i₂) : c'.Rel (f i₁) (f i₂)
+
+namespace Embedding
+
+variable {c c'}
+variable (e : Embedding c c')
+
+/-- An embedding of complex shapes `e` satisfies `e.IsRelIff` if the implication
+`e.rel` is an equivalence. -/
+class IsRelIff : Prop where
+  rel' (i₁ i₂ : ι) (h : c'.Rel (e.f i₁) (e.f i₂)) : c.Rel i₁ i₂
+
+lemma rel_iff [e.IsRelIff] (i₁ i₂ : ι) : c'.Rel (e.f i₁) (e.f i₂) ↔ c.Rel i₁ i₂ := by
+  constructor
+  · apply IsRelIff.rel'
+  · exact e.rel
+
+section
+
+variable (c c')
+variable (f : ι → ι') (hf : Function.Injective f)
+    (iff : ∀ (i₁ i₂ : ι), c.Rel i₁ i₂ ↔ c'.Rel (f i₁) (f i₂))
+
+/-- Constructor for embeddings between complex shapes when we have an equivalence
+`∀ (i₁ i₂ : ι), c.Rel i₁ i₂ ↔ c'.Rel (f i₁) (f i₂)`. -/
+@[simps]
+def mk' : Embedding c c' where
+  f := f
+  injective_f := hf
+  rel h := (iff _ _).1 h
+
+instance : (mk' c c' f hf iff).IsRelIff where
+  rel' _ _ h := (iff _ _).2 h
+
+end
+
+/-- The condition that the image of the map `e.f` of an embedding of
+complex shapes `e : Embedding c c'` is stable by `c'.next`. -/
+class IsTruncGE extends e.IsRelIff : Prop where
```

Comment:
The docstring is a bit unclear, since the definition doesn't mention `c'.next`. 

Reply:
The definition involves `c'.Rel`, and it is quite obvious that it is equivalent (because `c'.next i' = i'` when there is no `j'` such that `c'.Rel i' j'`). I am not so sure that the docstring would be better if it would just be a paraphrase of the definition.